spec/core/ics-004-channel-and-packet-semantics/README-pubsub.md

```diff
@@ -0,0 +1,274 @@
+    abortTransactionUnless(channel !== null)
+    abortTransactionUnless(channel.state === OPEN)
+
+    // check if the calling module owns the sending port
+    abortTransactionUnless(authenticateCapability(channelCapabilityPath(sourcePort, sourceChannel), capability))
+
+    // increment the send sequence counter
+    sequence = provableStore.get(nextSequenceSendPath(packet.sourcePort, packet.sourceChannel))
+    provableStore.set(nextSequenceSendPath(packet.sourcePort, packet.sourceChannel), sequence+1)
+
+    // store commitment to the packet data & packet timeout
+    provableStore.set(
+      packetCommitmentPath(sourcePort, sourceChannel, sequence),
+      hash(data)
+    )
+
+    // log that a packet can be safely sent
+    emitLogEntry("broadcastPacket", {sequence: sequence, data: data})
+}
+```
+
+#### Delivering packets
+
+The `deliverPacket` function is called by a module in order to receive an IBC broadcast packet sent on the corresponding broadcast channel end on the counterparty chain.
+
+The IBC handler performs the following steps in order:
+
+- Checks that the broadcast channel & connection are open to receive packets
+- Checks that the calling module owns the receiving port
+- Checks that the packet metadata matches the channel & connection information
+- Checks that the packet sequence is the next sequence the channel end expects to receive (for ordered)
+- Checks the inclusion proof of packet data commitment in the outgoing chain's state
+- Sets a store path to indicate that the packet has been received (for unordered channels only)
+- Increments the packet receive sequence associated with the channel end (ordered)
+
+We pass the address of the `relayer` that signed and submitted the packet to enable a module to optionally provide some rewards. This provides a foundation for fee payment, but can be used for other techniques as well (like calculating a leaderboard).
+
+```typescript
+function deliverPacket(
+  packet: OpaquePacket,
+  proof: CommitmentProof,
+  proofHeight: Height,
+  relayer: string): Packet {
+
+    channel = provableStore.get(channelPath(packet.destPort, packet.destChannel))
+    abortTransactionUnless(channel !== null)
+    abortTransactionUnless(channel.state === OPEN)
+    abortTransactionUnless(authenticateCapability(channelCapabilityPath(packet.destPort, packet.destChannel), capability))
+    abortTransactionUnless(packet.sourcePort === channel.counterpartyPortIdentifier)
+    abortTransactionUnless(packet.sourceChannel === channel.counterpartyChannelIdentifier)
+
+    connection = provableStore.get(connectionPath(channel.connectionHops[0]))
+
+    abortTransactionUnless(connection !== null)
+    abortTransactionUnless(connection.state === OPEN)
+
+    abortTransactionUnless(connection.verifyPacketData(
+      proofHeight,
+      proof,
+      packet.sourcePort,
+      packet.sourceChannel,
+      packet.sequence,
+      hash(packet.data)
+    ))
+
+    switch channel.order {
+      case ORDERED:
```

Comment:
I can see where the confusion comes from. I thought it is simpler to allows bcast channels to be either `ORDERED` or `UNORDERED`, as those are simpler terms than `ORDERED_ALLOW_TIMEOUT`. Nevertheless, I understand that semantically and ordered bcast channel is closer to an `ORDERED_ALLOW_TIMEOUT` normal channel.

Reply:
Think the binary choice of ORDERED and UNORDERED makes more sense. Since timeouts aren't even relevant here.

But maybe we can treat both ORDERED types the same way

Reply:
Ok, I think am confused by the fact that in `broadcastChanOpen` the order is checked to be `ORDERED_ALLOW_TIMEOUT` and the channel end in the subscriber will also be of order `ORDERED_ALLOW_TIMEOUT`, right? But then when checking the channel order in `deliverPacket` I see `ORDERED` and `UNORDERED`... Is this ok?